Modules & Effects

r-zander opened the pull request:
@Nullformed hab mal nen PR zwecks Review aufgemacht

Nullformed committed:
Implementation Modules, Operation Modes, multiple effects

Nullformed committed:
VFX: Fixed missing tag name, removed null check

Nullformed committed:
Fixed loading of modules & current operations.
Energy generation as effect.
Introduced 'title' to base data, setting 'name' automatically.
Introduced effect operators.
Cleaned up moduleOperations.
Effect description for module operations.

r-zander commented on index.html:
```diff
@@ -578,6 +578,7 @@ <h2 class="modal-title fs-4 fw-light">Welcome to Progress Station</h2>
 <script type="text/javascript" src="js/p5.Color.js"></script>
 <script type="text/javascript" src="js/utils.js"></script>
 <script type="text/javascript" src="js/gameEvents.js"></script>
+<script type="text/javascript" src="js/types.js"></script>
```
Bissl komisch dass es jetzt neben classes.js noch types.js gibt. Das klingt & sieht beides sehr ähnlich aus. Ich hätte nichs dagegen, types.js einfach in classes.js zu mergen.

r-zander commented on js/classes.js:
```diff
@@ -4,14 +4,19 @@ class Task {
         this.baseData = baseData;
-        this.name = baseData.name;
+        if (baseData.name !== undefined)
+            this.name = baseData.name;
```
das' relativ merkwürdig - entweder es gibt für alle Tasks in der baseData einen namen oder für keinen. So ein optionales, wichtiges Feld finde ich nicht gut

r-zander commented on js/classes.js:
```diff
@@ -20,6 +25,13 @@ class Task {
         return Math.round(this.getMaxXp() - this.xp);
     }
 
+    pushDefaultMultipliers(){
```
mir gefällt der Name nicht. Würde eher vorschlagen "prepareEffects" oder "collectEffects"

Nullformed replied:
Puh ja, das werden hauptsächlich XP oder Energy related modifier sein? Idk.

Nullformed committed:
Fixed getEnergyUsage if no usage defined.

Nullformed committed:
AutoPromote for modules WIP

r-zander commented on js/classes.js:
```diff
@@ -28,6 +40,35 @@ class Task {
+    }
+
+    getEffect(effectType) {
+        for (const id in this.baseData.effects){
```
You could have a map of effectType -> baseValue

Nullformed replied:
Gibt vermutlich ja eh nur 1 bis max. 3 effects?  Maps haben für mich wie gesagt bisher nie funktioniert. 🤷

r-zander commented on js/classes.js:
```diff
@@ -28,6 +40,35 @@ class Task {
+                return this.calculateEffectValue(effect);
+            }
+        }
+        return effectType.operator === 'x' ? 1 : 0;
```
~~why is the operator back? The calling side has to decide what to do ~~
Puh, wir müssen echt mal mit Janik reden ob nicht doch ausschließlich multiplikative Effekte okay sind. Sorry, ich hab mich da in was versteigert, was wohl die Komplexität einfach komplett durch die Decke hebt.

Nullformed replied:
Selbst im UI muss man halt wissen ob da + 30 Energy oder x 30 Energy steht.
Klar kannst dir auch conditions bauen und by effect sortieren...

r-zander commented on js/classes.js:
```diff
@@ -28,6 +40,35 @@ class Task {
+    }
+
+    getEffectDescription() {
+        let output = "";
+        for (let i = 0; i < this.baseData.effects.length; i++){
+            let effect = this.baseData.effects[i];
+            output += `${effect.effectType.operator}${this.calculateEffectValue(effect).toFixed(2)}`;
+            output += ` ${effect.effectType.description}`;
+            if (i < this.baseData.effects.length -1)
+            {
+                output += ', '
+            }
+        }
+
+        return output;
```
```suggestion
        return this.baseData.effects.map(function(effect) {
            return effect.effectType.operator +
                this.calculateEffectValue(effect).toFixed(2) +
                effect.effectType.description;
            }, this)
            .join(', ');
```

r-zander replied:
2 Sachen dazu: Bitte keine Template-Strings verwenden, wenn es literally nur concatenations sind aka die Syntax von string + string einfach weniger Zeichen.
Und ja, die fluide Array API ist perfekt für solche Aufgaben.

r-zander commented on js/classes.js:
```diff
@@ -60,12 +101,28 @@ class Job extends Task {
+        //this.xpMultipliers.push(getBoundItemEffect('Personal squire'));
+    }
+
+    do(){
+        super.do();
+    }
```
Kannst du weglassen, das macht nichts

r-zander commented on js/classes.js:
```diff
@@ -4,14 +4,19 @@ class Task {
     constructor(baseData) {
         this.type = this.constructor.name.toLowerCase();
         this.baseData = baseData;
```
Ich weiß, das war vorher schon so, aber ich wäre sehr dafür, dass alle Werte aus der BaseData ausgelesen und als properties gespeichert & verwendet werden, statt hier das ganze Objekt zu referenzieren und später einfach nach belieben daraus Felder zu verwenden.

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
     }
 
+    pushDefaultMultipliers(){
+        //this.expenseMultipliers.push(getBoundTaskEffect('Bargaining'));
```
super.pushDefaultMultipliers(); fehlt?

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+        for (let component in this.components){
+            component.setEnabled(this.#enabled)
+        }
+        if (this.#enabled){
```
```suggestion
        if (this.#enabled) {

```

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+            component.setEnabled(this.#enabled)
+        }
+        if (this.#enabled){
+            if (!gameData.currentModules.hasOwnProperty(this.name)){
```
tbf kannst du diesen Check weglassen. Ob `gameData.currentModules[this.name]` halt schon gesetzt ist oder nicht, ändert absolut nichts an den Auswirkungen der nächsten Zeile. Danach wir `gameData.currentModules[this.name]` gesetzt sein

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+                gameData.currentModules[this.name] = this;
+            }
+        }
+        else{
+            if (gameData.currentModules.hasOwnProperty(this.name)) {
```
Kann als `else if` verkürzt werden

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+    }
+}
+
+class ModuleComponent{
```
```suggestion
class ModuleComponent {

```
check mal bitte insgesamt dein Formatting bzw dass der IntelliJ Formatter angewendet wird

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+    }
+
+    setEnabled(value) {
+        if (this.currentMode !== null){
+            this.currentMode.setEnabled(value);
+        }
```
jo, das' jetzt halt merkwürdig - es gibt "currentMode" aber der kann enabled oder auch nicht sein 🤔

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+
+    //Support only one active mode
+    //Introduce default mode?
+    setActiveMode(modeId){
```
Das ist keine ModeId sondern eine reference auf den gewünschten, neuen active Mode. Also eher `newMode`

r-zander replied:
Kinda rel: Überall sonst sind das (Module)Operations --> bitte einheitlich benennen.

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+}
+
+class ModuleOperation extends Job{
+    enabled = false;
```
Sollte auch private sein

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+class ModuleOperation extends Job{
+    enabled = false;
+
+    do(){
+        super.do();
+    }
```
doesn't do anything

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+        }
+    }
+
+    toggleEnabled(){
```
Könnte man auch einfach nur "toggle" nennen :)

r-zander commented on js/classes.js:
```diff
@@ -91,24 +147,119 @@ class Item {
+    setEnabled(value) {
+        if (this.enabled === value) return;
+        this.enabled = value;
+        if (this.enabled){
+            if (!gameData.currentOperations.hasOwnProperty(this.name)){
+                gameData.currentOperations[this.name] = this;
+            }
+        }
+        else{
+            if (gameData.currentOperations.hasOwnProperty(this.name)) {
+                delete gameData.currentOperations[this.name];
+            }
+        }
```
Siehe Comments oben bzgl dieser Struktur

r-zander commented on js/gameConfig.js:
```diff
@@ -25,96 +25,74 @@ const units = {
-    'Blood meditation': {name: 'Blood meditation', maxXp: 100, effect: 0.01, description: 'Evil gain'},
-    'Demon\'s wealth': {name: 'Demon\'s wealth', maxXp: 100, effect: 0.002, description: 'Energy generation'},
-
+    Concentration: {name: 'Concentration', maxXp: 100, effects: [{effectType: EffectType.Population, baseValue: 5}, {effectType: EffectType.Energy, baseValue: 5}]},
```
```suggestion
    Concentration: {name: 'Concentration', maxXp: 100, effects: [{type: EffectType.Population, baseValue: 5}, {type: EffectType.Energy, baseValue: 5}]},
```
Einfach weil die ja schon Teil der effects sind 🙃

r-zander commented on js/gameConfig.js:
```diff
@@ -25,96 +25,74 @@ const units = {
+    Diesel: new ModuleOperation({title: 'Diesel', effects: [{effectType: EffectType.Population, baseValue: 5 }, {effectType: EffectType.Energy, baseValue: 5 }], maxXp: 50}),
+    Plastics: new ModuleOperation({title: 'Plastics', effects: [{effectType: EffectType.Population, baseValue: 5 }, {effectType: EffectType.Energy, baseValue: 5 }], maxXp: 100}),
+    Steel: new ModuleOperation({title: 'Steel', effects: [{effectType: EffectType.Population, baseValue: 5 }, {effectType: EffectType.Energy, baseValue: 5 }], maxXp: 200}),
+    QuantumReplicator: new ModuleOperation({title: 'Quantum Replicator', effects: [{effectType: EffectType.Population, baseValue: 5 }], maxXp: 400, energyConsumption: 3}),
```
Gönn dir mal hier noch eine Leerzeile um die Operations per Component zu gruppieren.

r-zander commented on js/gameConfig.js:
```diff
@@ -25,96 +25,74 @@ const units = {
+const moduleComponents = {
+    Fuel: new ModuleComponent({title: 'Fuel', operations: [moduleOperations.Garbage, moduleOperations.Diesel]}),
+    Products: new ModuleComponent({title: 'Products', operations: [moduleOperations.Plastics, moduleOperations.Steel]}),
+    Replication: new ModuleComponent({title: 'Replication', operations: [moduleOperations.QuantumReplicator, moduleOperations.BioGenesisChamber, moduleOperations.NanoFertilityDrones]}),
```
Hier auch: gruppieren via Leerzeile bitte

r-zander commented on js/gameConfig.js:
```diff
@@ -25,96 +25,74 @@ const units = {
+    Misc: [itemBaseData.Book]
 };
 
+//Initialize names
+for (const [key, module] of Object.entries(moduleOperations)) {
+    module.name = key;
+    module.baseData.name = key;
+}
+for (const [key, module] of Object.entries(moduleComponents)) {
+    module.name = key;
+}
+for (const [key, module] of Object.entries(modules)) {
+    module.name = key;
+}
+
```
a) das ist nicht Teil der gameConfig, sondern sollte iwo nach Main 
b) ich denke nicht, dass du die baseData anpassen musst. Die wird ja theoretisch anfangs einmal eingelesen und dann "vergessen", kein Grund da den Namen zu setzen.
c) name --> identifier

r-zander commented on js/gameConfig.js:
```diff
@@ -130,67 +108,36 @@ const headerRowColors = {
-    'Study desk': 'A dedicated area which provides many fine stationary and equipment designed for furthering your progress in research.',
-    'Library': 'Stores a collection of books, each containing vast amounts of information from basic life skills to complex magic spells.',
+
+    "Quantum Replicator": "Introducing the 'Quantum Replicator'—the ultimate solution for population growth! This futuristic device uses quantum technology to duplicate individuals, allowing you to rapidly expand your population. With each activation, watch as your society flourishes and thrives. Just remember to keep track of the originals, or you might end up with an army of duplicates!",
```
Bitte in JS nach Möglichkeit single quotes verwenden

r-zander commented on js/gameConfig.js:
```diff
@@ -298,46 +253,17 @@ function addMultipliers() {
-        } else if (task instanceof Skill) {
-            task.xpMultipliers.push(getBoundTaskEffect('Concentration'));
-            task.xpMultipliers.push(getBoundItemEffect('Book'));
-            task.xpMultipliers.push(getBoundItemEffect('Study desk'));
-            task.xpMultipliers.push(getBoundItemEffect('Library'));
+            task.energyGenerationMultipliers = [];
```
die zwei Zeilen gehören wohl in den Task constructor

r-zander commented on js/gameConfig.js:
```diff
@@ -298,46 +253,17 @@ function addMultipliers() {
 
     for (let itemName in gameData.itemData) {
         const item = gameData.itemData[itemName];
         item.expenseMultipliers = [];
```
--> constructor

r-zander commented on js/main.js:
```diff
@@ -80,47 +71,37 @@ function applyMultipliers(value, multipliers) {
+function getPopulation() {
+    let population = 0;
+    const tasks = gameData.currentOperations;
+    if (tasks !== null){
```
die sind doch auf ein leeres Object initialisiert und ich hoffe mal niemand nullt dieses Feld

r-zander commented on js/main.js:
```diff
@@ -80,47 +71,37 @@ function applyMultipliers(value, multipliers) {
+        for (const taskName in tasks){
+            const task = tasks[taskName];
+            if(task != null)
+                population += task.getEffect(EffectType.Population);
```
ja nice, find ich nicht verkehrt.

(bitte noch { ... } außenrum)

r-zander commented on js/main.js:
```diff
@@ -80,47 +71,37 @@ function applyMultipliers(value, multipliers) {
-    return baseGameSpeed * +isPlaying() * timeWarpingSpeed;
+    //const timeWarping = gameData.taskData['Time warping'];
+    //const timeWarpingSpeed = gameData.timeWarpingEnabled ? timeWarping.getEffect() : 1;
+    return baseGameSpeed * +isPlaying();
```
```suggestion
    return baseGameSpeed * +isPlaying() /* * timeWarpingSpeed */;

```

r-zander commented on js/main.js:
```diff
@@ -163,7 +144,7 @@ function setTask(taskName) {
     const task = gameData.taskData[taskName];
     if (task instanceof Job) {
-        gameData.currentJob = task;
+        console.error("setTask with job no longer supported")
```
```suggestion
        throw new Error("setTask with job no longer supported");
```

r-zander commented on js/main.js:
```diff
@@ -226,6 +207,114 @@ function createRequiredRow(categoryName) {
+        level4DomGetter.byClass('name').textContent = mode.title;
+        let descriptionElement = level4DomGetter.byClass('descriptionTooltip');
+        descriptionElement.ariaLabel = mode.title;
+        descriptionElement.title = tooltips[mode.title];
```
das wäre ein gutes Argument dafür, die Tooltips auch direkt an die Objekte anzuhängen statt sie extra rauszusuchen. Via .title sollte überhaupt nichts lookupped werden, höchstens via .identifier

r-zander commented on js/main.js:
```diff
@@ -226,6 +207,114 @@ function createRequiredRow(categoryName) {
+            //level4DomGetter.byClass('progressBar').onclick = mode.toggleEnabled.bind(mode);
+            level4DomGetter.byClass('progressBar').onclick = function () {
+                module.setActiveMode(mode);
+            }
```
```suggestion
            };
```

r-zander commented on js/main.js:
```diff
@@ -226,6 +207,114 @@ function createRequiredRow(categoryName) {
+    const level4Elements = [];
+    const operations = module.operations;
+
+    operations.forEach(function (mode) {
```
mode --> operation

r-zander commented on js/main.js:
```diff
@@ -226,6 +207,114 @@ function createRequiredRow(categoryName) {
+function createLevel3Elements(categoryName, module, level3Slot, domId) {
+    const level3Elements = [];
+
+    for (let comp of module.components) {
```
```suggestion
    for (const component of module.components) {
```

r-zander commented on js/main.js:
```diff
@@ -226,6 +207,114 @@ function createRequiredRow(categoryName) {
+    slot.replaceWith(...level1Elements);
+}
+
+//// TODO: Remove obsolete func
```
Joa, sollte weg ^^

r-zander commented on js/main.js:
```diff
@@ -226,6 +207,114 @@ function createRequiredRow(categoryName) {
+    for (let module of category) {
+        const level2Element = Dom.new.fromTemplate('level2Template');
+        const level2DomGetter = Dom.get(level2Element);
+        level2DomGetter.byClass('name').textContent = module.title;
```
wäre schön wenn die CSS Class dann auch unbenannt wird

r-zander commented on js/main.js:
```diff
@@ -470,7 +558,10 @@ function updateTaskRows() {
-        if (task === gameData.currentJob || task === gameData.currentSkill) {
+        if (task instanceof ModuleOperation && gameData.currentOperations.hasOwnProperty(task.name)) {
+            progressFill.classList.add('current');
+        }
+        if (task instanceof Skill && task === gameData.currentSkill) {
```
```suggestion
        } else if (task instanceof Skill && task === gameData.currentSkill) {
```

r-zander commented on js/main.js:
```diff
@@ -639,18 +733,45 @@ function updateBodyClasses() {
+    }*/
+
+    const tasks = gameData.currentOperations;
+    if (tasks !== null){
```
should never be null

r-zander commented on js/main.js:
```diff
@@ -639,18 +733,45 @@ function updateBodyClasses() {
+    if (tasks !== null){
+        for (const taskName in tasks){
+            const task = tasks[taskName];
+            if(task != null)
```
Hoffentlich ist das nie null 😅 also, sollte einfach nicht

r-zander commented on js/main.js:
```diff
@@ -639,18 +733,45 @@ function updateBodyClasses() {
+
+    const tasks = gameData.currentOperations;
+    if (tasks !== null){
+        for (const taskName in tasks){
+            const task = tasks[taskName];
+            if(task != null)
+                task.do();
+        }
+    }
```
_Könnte_ man auch einfach so bauen:
```suggestion
        tasks.forEach(function (task) { task.do(); });
```
oder richtig dope mit ES6:
```
        tasks.forEach(task => task.do());
```
wenn wir das wollen sollten wir uns mal über unsere JavaScript Version unterhalten. Browser support sollte vollkommen okay sein für ES6

r-zander commented on js/main.js:
```diff
@@ -639,18 +733,45 @@ function updateBodyClasses() {
+
 function doTask(task) {
     if (task == null) return;
     if (task instanceof LayeredTask && task.isDone()) return;
```
das checkt der LayeredTask doch intern eh selbst

r-zander commented on js/main.js:
```diff
@@ -659,9 +780,39 @@ function getMaxEnergy() {
+                energyUsage += task.getEnergyUsage();
+        }
+    }
+    energyUsage += gameData.currentProperty.getEnergyUsage();
```
das ist schon in Zeile 785 passiert

r-zander commented on js/main.js:
```diff
@@ -980,9 +1144,9 @@ function isPlaying() {
-        if (task.baseData.energyGeneration) {
-            task.baseData = jobBaseData[task.name];
-            task = Object.assign(new Job(jobBaseData[task.name]), task);
+        if (task instanceof Job) {
```
ist hier nicht genau das Problem, dass die deserialisierten Daten keine instances mehr sind sondern nur noch objects?

Nullformed replied:
Scheinbar nicht. Dachte ich auch

Nullformed replied:
Ach so, ja das passiert jetzt vorher im replaceSaveDict...

r-zander commented on js/main.js:
```diff
@@ -1048,7 +1211,14 @@ function replaceSaveDict(dict, saveDict) {
             saveDict[key] = dict[key];
-        } else if (dict === gameData.requirements) {
+        }
+        if (dict === gameData.taskData && dict[key] instanceof Job) {
+            dict[key].level = saveDict[key].level;
+            dict[key].maxLevel = saveDict[key].maxLevel;
+            dict[key].xp = saveDict[key].xp;
+            saveDict[key] = dict[key];
+        }
```
Keine Ahnung was das macht oder ob es klappt

Nullformed replied:
Siehe https://github.com/r-zander/progress-station/pull/5#discussion_r1339296137.

Nullformed replied:
Sollte man wohl noch vereinheitlichen.

r-zander commented on js/main.js:
```diff
@@ -1076,6 +1246,13 @@ function loadGameData() {
         replaceSaveDict(gameData.itemData, gameDataSave.itemData);
         replaceSaveDict(gameData.battleData, gameDataSave.battleData);
 
+        for (let id in gameDataSave.currentOperations){
+            gameDataSave.currentOperations[id] = moduleOperations[id];
+        }
+        for (let id in gameDataSave.currentModules){
+            gameDataSave.currentModules[id] = modules[id];
+        }
```
Fehlen da nicht die eigentlichen Spielwerte dann in gameDataSave?

Nullformed replied:
Ne das sind ja nur die 'current'. Das könnten auch einfach Ids sein sind aber gerade halt Referenzen

Nullformed replied:
Aber das interne leben der Module wird hier noch nicht wiederhergestellt (on/off… current mode).

r-zander commented on js/main.js:
```diff
@@ -1244,20 +1419,24 @@ function displayLoaded() {
 
-    createData(gameData.taskData, jobBaseData);
+    //direct ref assignment - taskData could be replaced
+    for (let entityData of Object.values(moduleOperations)) {
+        entityData.id = 'row ' + entityData.name;
+        gameData.taskData[entityData.name] = entityData
+    }
```
Warum passiert das hier so "nackt" mitten in init()?

Nullformed replied:
🤷 Wenn man saves überarbeitet könnte die Taskdata auch weg.

r-zander commented on js/vfx.js:
```diff
@@ -186,7 +186,8 @@ GameEvents.TaskLevelChanged.subscribe(function (taskInfo) {
     }
     else{
-        taskProgressBar = getTaskElement(taskInfo.name).querySelector('.progressBar');
+        const taskElement = getTaskElement(taskInfo.name);
```
Könnte auch eine entspannte Methode in Task sein `task.getElement()`

Nullformed committed:
Code review cleanup.
Task calls changed to module.do().
Toggling of Modules via button.
Migrate 'current' modules before replacing save data.

Nullformed committed:
Code review cleanup. https://github.com/r-zander/progress-station/pull/5#discussion_r1339215944

Nullformed committed:
Create rows cleanup.
